core/mcp.go

```diff
@@ -633,103 +747,202 @@ Each parameter corresponds to a named result with a specific purpose. Do not cal
 			if !ok {
 				return nil, fmt.Errorf("invalid arguments: %T", args)
 			}
-			for name := range m.env.outputsByName {
+			for name, output := range m.env.outputsByName {
 				arg, ok := vals[name]
 				if !ok {
-					return nil, fmt.Errorf("required argument %s not provided", name)
+					return nil, fmt.Errorf("required output %s not provided", name)
 				}
 				argStr, ok := arg.(string)
 				if !ok {
 					return nil, fmt.Errorf("invalid type for argument %s: %T", name, arg)
 				}
-				bnd, ok := m.env.objsByID[argStr]
-				if !ok {
-					return nil, fmt.Errorf("object not found for argument %s: %s", name, argStr)
-				}
-				obj := bnd.Value
-				if output, found := m.env.Output(name); found {
-					if expected := output.expectedType; expected != nil {
-						expectedType := expected.TypeName()
-						actualType := obj.Type().Name()
-						if expectedType != actualType {
-							return nil, fmt.Errorf("incompatible types: %s must be %s, got %s", name, expectedType, actualType)
-						}
+				if output.ExpectedType == "String" {
+					output.Value = dagql.String(argStr)
+				} else {
+					bnd, ok := m.env.objsByID[argStr]
+					if !ok {
+						return nil, fmt.Errorf("object not found for argument %s: %s", name, argStr)
+					}
+
+					obj := bnd.Value
+
+					// Propagate description from output to binding so that outputs are
+					// described under `Available objects:`
+					bnd.Description = output.Description
+
+					actualType := obj.Type().Name()
+					if output.ExpectedType != actualType {
+						return nil, fmt.Errorf("incompatible types: %s must be %s, got %s", name, output.ExpectedType, actualType)
 					}
 					output.Value = obj
-				} else {
-					return nil, fmt.Errorf("undefined output: %q", name)
 				}
 			}
 			m.returned = true
 			return "ok", nil
 		},
-	}
+	}, true
 }
 
-func (m *MCP) Builtins(srv *dagql.Server) ([]LLMTool, error) {
-	builtins := []LLMTool{}
-
-	if len(m.env.outputsByName) > 0 {
-		builtins = append(builtins, m.returnBuiltin())
+func (m *MCP) Builtins(srv *dagql.Server, tools []LLMTool) ([]LLMTool, error) {
+	builtins := []LLMTool{
+		{
+			Name:        "think",
```

Comment:
On the fence here - it's very interesting to see what it writes, and it's more likely to follow the plan that it writes. We can also give it custom UI. For now it just has a :thought_balloon: emoji but I'd ilke to dim the text.

But I haven't validated whether it actually helps the evals, in isolation from all the other changes.